### PR TITLE
fix: avoid `export *` in magic files for shaking

### PR DESCRIPTION
**Pull request by jacob-ebey**
fix: add mainFields to client build to resolve ESM
fix: force sideEffects false for browser route modules

### CODE REVIEW

**Comment by jacob-ebey** [comment]
Fixes https://github.com/remix-run/remix/issues/1618

File: packages/remix-dev/compiler.ts [hunk 1]
```diff
@@ -356,6 +356,7 @@ async function createBrowserBuild(
     sourcemap: options.sourcemap,
     metafile: true,
     incremental: options.incremental,
+    mainFields: ["module", "browser", "main"],
```
Comment (jacob-ebey) [inline packages/remix-dev/compiler.ts:359]: I think browser might want to come first on second thought.

File: packages/remix/index.ts [hunk 2]
```diff
@@ -1,6 +1 @@
-// @ts-ignore
-export * from "./client";
-// @ts-ignore
-export * from "./server";
-// @ts-ignore
-export * from "./platform";
+throw new Error("Did you forget to run `remix setup` for your platform?");
```
Comment (kentcdodds) [inline packages/remix/index.ts:1]: Oh nice! This will be very helpful :)

File: packages/remix-dev/compiler.ts (hunk 1 above)
Comment (mjackson) [inline packages/remix-dev/compiler.ts:359]: This is the main line that makes me a little nervous about this PR. If we're specifically building for the browser, and someone specifies the browser field in package.json, it feels like that should take precedence. Can you say a little more about why this change is necessary?

File: packages/remix-dev/compiler/plugins/browserRouteModulesPlugin.ts [hunk 3]
```diff
@@ -38,7 +38,11 @@ export function browserRouteModulesPlugin(
       );
 
       build.onResolve({ filter: suffixMatcher }, args => {
-        return { path: args.path, namespace: "browser-route-module" };
+        return {
+          path: args.path,
+          namespace: "browser-route-module",
+          sideEffects: false
```
Comment (mjackson) [inline packages/remix-dev/compiler/plugins/browserRouteModulesPlugin.ts:44]: Why do we need this?

File: packages/remix-dev/compiler.ts (hunk 1 above)
Comment (jacob-ebey) [reply packages/remix-dev/compiler.ts]: My thought was that the "browser" field just means "this can run in the browser" and doesn't mean "ESM format". Since we are bundling to ESM, might as well take the module if it exists. On second thought though I've reverted it to be browser first. We need some definition for mainFields here though otherwise esbuild falls back to it's default behavior of "if something was `required` anywhere, always take "main". Without this we break client bundle tree shaking.

File: packages/remix-dev/compiler/plugins/browserRouteModulesPlugin.ts (hunk 3 above)
Comment (jacob-ebey) [reply packages/remix-dev/compiler/plugins/browserRouteModulesPlugin.ts]: We don't, leftover code that I removed :)
